File: hackerrank/problem solving/algorithms/2 - implementation/40 - acm icpc team/with morfo library/acm_icpc_team.js
// https://www.hackerrank.com/challenges/acm-icpc-team/problem?isFullScreen=true

import { println, read, skipInput } from 'morfo';

function main() {
  const n = read(Number);
  skipInput();
  const binaries = readBinaries(n);
  println(acmTeam(binaries));
}

function readBinaries(n) {
  return read(n, String);
}

function acmTeam(binaries) {
  let maxSubjects = 0;
  let teamsWithMax = 0;

  for (let i = 0, n1 = binaries.length - 1; i < n1; i++) {
    for (let j = i + 1; j < binaries.length; j++) {
      const knownSubjects = countSubjectsKnownBy2Teams(
        binaries[i],
        binaries[j],
      );

      if (knownSubjects > maxSubjects) {
        maxSubjects = knownSubjects;
        teamsWithMax = 1;
      } else if (knownSubjects === maxSubjects) teamsWithMax++;
    }
  }

  return [maxSubjects, teamsWithMax];
}

function countSubjectsKnownBy2Teams(a, b) {
  let count = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] === '1' || b[i] === '1') count++;
  }
  return count;
}

main();

File: hackerrank/problem solving/algorithms/2 - implementation/40 - acm icpc team/with morfo library/acm_icpc_team.go
// https://www.hackerrank.com/challenges/acm-icpc-team/problem?isFullScreen=true

package main

import (
    "morfo/morfoio"
)

func main() {
    n := morfoio.Read[int]()
    morfoio.SkipInput()
    binaries := readBinaries(n)
    maxSubjects, teamsWithMax := acmTeam(binaries)
    morfoio.Println([]int{maxSubjects, teamsWithMax})
}

func readBinaries(n int) []string {
    return morfoio.ReadN[string](n)
}

func acmTeam(binaries []string) (int, int) {
    maxSubjects := 0
    teamsWithMax := 0

    for i, a := range binaries[:len(binaries)-1] {
        for _, b := range binaries[i+1:] {
            knownSubjects := countSubjectsKnownBy2Teams(a, b)

            if knownSubjects > maxSubjects {
                maxSubjects = knownSubjects
                teamsWithMax = 1
            } else if knownSubjects == maxSubjects {
                teamsWithMax++
            }
        }
    }

    return maxSubjects, teamsWithMax
}

func countSubjectsKnownBy2Teams(a, b string) int {
    count := 0
    for i := range a {
        if a[i] == '1' || b[i] == '1' {
            count++
        }
    }
    return count
}
